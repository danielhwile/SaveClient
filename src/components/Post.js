import { BiMessageDetail, BiEdit, BiTrash} from "react-icons/bi";
import { FaArrowDown, FaArrowUp } from "react-icons/fa";
import { useState } from 'react';
import { gql, useMutation } from '@apollo/client';
import { UserContext } from "../UserContext";
import { useContext } from 'react';
import { IconContext } from "react-icons";
// local imports
import { timeCalculator, longDateTimeString } from "../utils";
import '../static/Post.css'

// graphQL query/mutation for sending votes
const VOTE_PUT = gql`
    mutation putVote($username: String!, $post_id: String!, $vote: Int!) {
        putVote(username: $username, post_id:$post_id, vote:$vote) {
            vote
        }
    } 
`;

const Post = ({postDetails, isCommentSection}) => {
    // component for individual posts, both at the top of their comments section and on post query pages

    // setting vote states
    const [voteValue, setVoteValue] = useState(postDetails.userVote.vote);
    const [voteSum, setVoteSum] = useState(postDetails.voteSum-postDetails.userVote.vote);

    //logged in user is needed for applying their votes to the right person
    const viewerUser = useContext(UserContext);

    const [putVote] = useMutation(VOTE_PUT); // fx to send the vote mutation in the upvote and downvote fxs

    const upvoteFunc = (e) => {
            //sends an upvote to the server from the logged in user on click
            setVoteValue(1);
            putVote({ variables: {username:viewerUser.username,post_id:postDetails._id,vote:1}});
    }
    const downvoteFunc = (e) => {
            //sends a downvote to the server from the logged in user on click
            setVoteValue(-1);
            putVote({ variables: {username:viewerUser.username,post_id:postDetails._id,vote:-1}});
    }

    const userProfileClick = (e) => {
        // link to user profile at the username in the post header
        e.preventDefault();
        let hrefString ='/user/'+postDetails.creator;
        window.location.href=hrefString;
    }

    const postCommentClick = (e) => {
        // link to the post's comment section.
        e.preventDefault();
        let hrefString ='/post/'+postDetails._id;
        window.location.href=hrefString;
    }
    return (
        <div className="postDiv">
            <div className="votesDiv">
                <label class="labelVote">
                    <input type="radio" className="voteRadio voteRadioUp" name="vote" value={voteValue} checked={(voteValue == 1)} onChange={upvoteFunc}/>
                    <h2 className={(voteValue === 1) && "colorUpVote"}><FaArrowUp/></h2>
                </label>
                <h3>{voteSum + voteValue}</h3> {/*Rather than callback the current votesum and have confusing vote total changes (If other users voted simultaneously, we adjust the total locally*/}
                <label class="labelVote">
                    <input type="radio" className="voteRadio" name="vote" value={voteValue} checked={(voteValue == -1)} onChange={downvoteFunc}/>
                    <h2 className={(voteValue === -1) && "colorDownVote"}><FaArrowDown/></h2>
                </label>
            </div>
            <div className="postContent">
                <div className="postHeaderandTitle">
                    <div className="postHeader">
                        <p>Created by <a className="pointer" onClick={userProfileClick}>{postDetails.creator}</a></p>
                        <p title={longDateTimeString(postDetails.createdAt)}>{timeCalculator(postDetails.createdAt)}</p>
                    </div>
                    <h2 className="postTitle">{postDetails.title}</h2>
                    {isCommentSection&& /*If we are in the comment section we hide the comment section link and insert the description */
                        <p className="postDescription">{postDetails.description}</p>}
                </div>
                <div className="postFooter">
                    {(!isCommentSection)&& /* If we are not in the comment section we display the link to the comment section*/
                        <h3 className="postComment" onClick={postCommentClick}>
                            <IconContext.Provider value={{ style: { verticalAlign: 'middle'} }}>
                                <BiMessageDetail/>
                            </IconContext.Provider>
                            Comments
                        </h3>}
                    {/*NOTE: ADD LATER FOR PORTFOLIO VERSION NOT OUTLINED IN PORJECT <h3><BiEdit/></h3>  <h3><BiTrash/></h3>*/}
                </div>
            </div>
        </div>
    )
}

export default Post
