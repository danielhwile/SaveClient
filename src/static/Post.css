.voteRadio {
  display: none !important;
}
   
.colorUpVote {
  color: rgb(29, 150, 25);
}
.colorDownVote {
  color: rgb(190, 65, 27);
}
.postDiv {
  border: 2px rgb(170, 170, 170) solid;
  background-color: whitesmoke;
  border-radius: 5px;
  padding: .5em;
  display: grid;
  grid-template-columns: 8% auto;
  margin: .6em;
  width: 90vw;
  max-width: 600px;
  min-width: 400px;
}
.postFooter{
  display: flex;
}
.postComment{
  background-color:rgb(226, 226, 226);
  border-radius: 3px;
  padding: .1em .3em;
  cursor: pointer;
}
.postComment:hover{
  background-color:rgb(208, 215, 231);
}
.postTitle{
  padding: .3em 0;
}
.postHeader{
  display: flex;
  justify-content: space-between;
}
.votesDiv{
  display: flex;
  align-items: center;
  flex-direction: column;
  justify-content: flex-start;
  padding: 0 .3em 0 0;
}
.labelVote{
  cursor: pointer;
}
.pointer{
  cursor: pointer;
}