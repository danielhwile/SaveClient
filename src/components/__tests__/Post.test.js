import React from 'react';
import { mount, render, shallow } from 'enzyme';
import Post from '../Post.js';
import { ApolloProvider, ApolloClient, InMemoryCache, createHttpLink} from '@apollo/client';
import { setContext } from '@apollo/client/link/context';
import jwt from 'jsonwebtoken';
import { IconContext } from "react-icons";
import { BiMessageDetail } from "react-icons/bi";
import renderer from 'react-test-renderer';

// All of below was needed to mount a component that works with Apollo Client

const httpLink = createHttpLink({
  uri: 'http://localhost:4000/api/graphql',
});

const token = localStorage.getItem('authToken');

const authLink = setContext((_, { headers }) => {
  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : "",
    }
  }
});

var user = null;

if (token) {
const tokenDecode = jwt.decode(token)
user = {username:tokenDecode.username};
}
const client = new ApolloClient({
  link: authLink.concat(httpLink),
  cache: new InMemoryCache()
});


// On to testing!

const examplePostProps = {title:"test Title", description:"test Description 401042140",createdAt:1632236399125,creator:"JestTest",_id:"a161",voteSum:"105",userVote:{vote:0}};
describe('Post', () => {
    it('should render in Front Page Mode (No Description)', () => {
      const component = mount(<ApolloProvider client={client}><Post postDetails={examplePostProps}/></ApolloProvider>);
      
      const creatorTag = <p title='Tuesday, September 21, 2021 at 10:59:59 AM EDT'>&#60;1 Hour Ago</p>;
      expect(component.contains(creatorTag)).toEqual(true);

      const descriptionText = "test Description 401042140";
      expect(component.contains(descriptionText)).toEqual(false);
      component.unmount();
    })
    it('should render in Comments Page Mode (With Description)', () => {
        const component = mount(<ApolloProvider client={client}><Post postDetails={examplePostProps} isCommentSection={true}/></ApolloProvider>);
        
        const creatorTag = <p title='Tuesday, September 21, 2021 at 10:59:59 AM EDT'>&#60;1 Hour Ago</p>;
        expect(component.contains(creatorTag)).toEqual(true);
  
        const descriptionText = "test Description 401042140";
        expect(component.contains(descriptionText)).toEqual(true);
        component.unmount();
      })
    it('should match snapShot', () => {
        const snap = renderer.create(<ApolloProvider client={client}><Post postDetails={examplePostProps} isCommentSection={true}/></ApolloProvider>).toJSON();
        expect(snap).toMatchSnapshot();
    })
});