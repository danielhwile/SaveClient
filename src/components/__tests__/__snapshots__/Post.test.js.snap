// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post should match snapShot 1`] = `
<div
  className="postDiv"
>
  <div
    className="votesDiv"
  >
    <label
      class="labelVote"
    >
      <input
        checked={false}
        className="voteRadio voteRadioUp"
        name="vote"
        onChange={[Function]}
        type="radio"
        value={0}
      />
      <h2
        className={false}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M34.9 289.5l-22.2-22.2c-9.4-9.4-9.4-24.6 0-33.9L207 39c9.4-9.4 24.6-9.4 33.9 0l194.3 194.3c9.4 9.4 9.4 24.6 0 33.9L413 289.4c-9.5 9.5-25 9.3-34.3-.4L264 168.6V456c0 13.3-10.7 24-24 24h-32c-13.3 0-24-10.7-24-24V168.6L69.2 289.1c-9.3 9.8-24.8 10-34.3.4z"
          />
        </svg>
      </h2>
    </label>
    <h3>
      105
    </h3>
    <label
      class="labelVote"
    >
      <input
        checked={false}
        className="voteRadio"
        name="vote"
        onChange={[Function]}
        type="radio"
        value={0}
      />
      <h2
        className={false}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
          />
        </svg>
      </h2>
    </label>
  </div>
  <div
    className="postContent"
  >
    <div
      className="postHeaderandTitle"
    >
      <div
        className="postHeader"
      >
        <p>
          Created by 
          <a
            className="pointer"
            onClick={[Function]}
          >
            JestTest
          </a>
        </p>
        <p
          title="Tuesday, September 21, 2021 at 10:59:59 AM EDT"
        >
          &lt;1 Hour Ago
        </p>
      </div>
      <h2
        className="postTitle"
      >
        test Title
      </h2>
      <p
        className="postDescription"
      >
        test Description 401042140
      </p>
    </div>
    <div
      className="postFooter"
    />
  </div>
</div>
`;
